parachain/pallets/commitments/src/lib.rs

```diff
@@ -0,0 +1,103 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use sp_std::prelude::*;
+use frame_support::{
+	decl_module, decl_storage, decl_event, decl_error,
+	weights::Weight,
+	traits::Get
+};
+
+use sp_io::hashing::keccak_256;
+use sp_core::{H160, H256};
+use sp_runtime::{
+	traits::Zero,
+	DigestItem
+};
+
+use codec::{Encode, Decode};
+use artemis_core::Commitments;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+/// Custom DigestItem for header digest
+#[derive(Encode, Decode, Copy, Clone)]
+enum CustomDigestItem {
+	Commitment(H256)
+}
+
+impl<T> Into<DigestItem<T>> for CustomDigestItem {
+    fn into(self) -> DigestItem<T> {
+        DigestItem::Other(self.encode())
+    }
+}
+
+#[derive(Encode, Decode, Clone)]
+struct Message {
+	address: H160,
+	payload: Vec<u8>,
+}
+
+pub trait Trait: frame_system::Trait {
+	type Event: From<Event> + Into<<Self as frame_system::Trait>::Event>;
+
+	type CommitInterval: Get<Self::BlockNumber>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Commitments {
+		/// messages for an application
+		Messages: Vec<Message>
+	}
+}
+
+decl_event! {
+	pub enum Event {}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		// Generate a message commitment every `T::PruneInterval` blocks
+		fn on_initialize(now: T::BlockNumber) -> Weight {
```

Comment:
If we generate a commitment in `on_finalize` for block N, we won't be able to prune storage at the same time. This is because the relayer will need to be able to query for messages produced during block N. So we'd be limited to pruning storage in `on_initialize` in block N + 1, which complicates things a bit.

Reply:
That makes sense. Let's keep it on `on_finalize` for now then.